proxy not work for post

POST request body is not proxied to the servers. Maybe the same problem as https://github.com/chimurai/http-proxy-middleware/issues/40

- Web Browser: chrome 86
- Local OS: macOS 10.13
- Remote OS: docker
- `code-server --version`:  3.7.4

1. start a server in code-server terminal
2. post to server without header 'application/json' by proxy, the server receive body
3. post to server with header 'application/json' by proxy, the server can not receive body

I managed to fix the issue (I think) but the yarn watch thing gives me a `[tsc] src/node/routes/pathProxy.ts:38:13 - error TS2339: Property 'body' does not exist on type 'IncomingMessage'.` for the 3 times i used `req.body`. It did seem to receive the body when i tested it with a simple express server despite this. I only did it for the subfolder proxies for now because they were easier to test and I had to learn what express is to do this so it might not be perfect

My edited src/node/routes/pathProxy.ts is [here](https://github.com/BenBJD/code-server/blob/master/src/node/routes/pathProxy.ts)

--- a/src/node/proxy.ts
+++ b/src/node/proxy.ts
@@ -9,6 +9,8 @@ proxy.on("error", (error, _, res) => {
 })
 
 // Intercept the response to rewrite absolute redirects against the base path.
+// Is disabled when the request has no base path which means --proxy-path-passthrough has
+// been enabled.
 proxy.on("proxyRes", (res, req) => {
   if (res.headers.location && res.headers.location.startsWith("/") && (req as any).base) {
     res.headers.location = (req as any).base + res.headers.location

--- a/src/node/routes/index.ts
+++ b/src/node/routes/index.ts
@@ -65,9 +65,6 @@ export const register = async (
   app.use(cookieParser())
   wsApp.use(cookieParser())
 
-  app.use(bodyParser.json())
-  app.use(bodyParser.urlencoded({ extended: true }))
-
   const common: express.RequestHandler = (req, _, next) => {
     // /healthz|/healthz/ needs to be excluded otherwise health checks will make
     // it look like code-server is always in use.
@@ -106,6 +103,12 @@ export const register = async (
   app.use("/", domainProxy.router)
   wsApp.use("/", domainProxy.wsRouter.router)
 
+  app.use("/proxy", proxy.router)
+  wsApp.use("/proxy", proxy.wsRouter.router)
+
+  app.use(bodyParser.json())
+  app.use(bodyParser.urlencoded({ extended: true }))
+
   app.use("/", vscode.router)
   wsApp.use("/", vscode.wsRouter.router)
   app.use("/vscode", vscode.router)
@@ -120,9 +123,6 @@ export const register = async (
       redirect(req, res, "/", {})
     })
   }
-
-  app.use("/proxy", proxy.router)
-  wsApp.use("/proxy", proxy.wsRouter.router)
 
   app.use("/static", _static.router)
   app.use("/update", update.router)

--- a/src/node/routes/pathProxy.ts
+++ b/src/node/routes/pathProxy.ts
@@ -28,8 +28,10 @@ router.all("/(:port)(/*)?", (req, res) => {
     throw new HttpError("Unauthorized", HttpCode.Unauthorized)
   }
 
-  // Absolute redirects need to be based on the subpath when rewriting.
-  ;(req as any).base = `${req.baseUrl}/${req.params.port}`
+  if (!req.args["proxy-path-passthrough"]) {
+    // Absolute redirects need to be based on the subpath when rewriting.
+    ;(req as any).base = `${req.baseUrl}/${req.params.port}`
+  }
 
   proxy.web(req, res, {
     ignorePath: true,

--- a/test/proxy.test.ts
+++ b/test/proxy.test.ts
@@ -1,26 +1,27 @@
+import bodyParser from "body-parser"
 import * as express from "express"
 import * as httpserver from "./httpserver"
 import * as integration from "./integration"
 
 describe("proxy", () => {
-  let codeServer: httpserver.HttpServer | undefined
   const nhooyrDevServer = new httpserver.HttpServer()
+  let codeServer: httpserver.HttpServer | undefined
   let proxyPath: string
+  let e: express.Express
 
   beforeAll(async () => {
-    const e = express.default()
-    await nhooyrDevServer.listen(e)
-    e.get("/wsup", (req, res) => {
-      res.json("asher is the best")
+    await nhooyrDevServer.listen((req, res) => {
+      e(req, res)
     })
     proxyPath = `/proxy/${nhooyrDevServer.port()}/wsup`
-    e.get(proxyPath, (req, res) => {
-      res.json("joe is the best")
-    })
   })
 
   afterAll(async () => {
     await nhooyrDevServer.close()
+  })
+
+  beforeEach(() => {
+    e = express.default()
   })
 
   afterEach(async () => {
@@ -31,6 +32,9 @@ describe("proxy", () => {
   })
 
   it("should rewrite the base path", async () => {
+    e.get("/wsup", (req, res) => {
+      res.json("asher is the best")
+    })
     ;[, , codeServer] = await integration.setup(["--auth=none"], "")
     const resp = await codeServer.fetch(proxyPath)
     expect(resp.status).toBe(200)
@@ -39,10 +43,61 @@ describe("proxy", () => {
   })
 
   it("should not rewrite the base path", async () => {
+    e.get(proxyPath, (req, res) => {
+      res.json("joe is the best")
+    })
     ;[, , codeServer] = await integration.setup(["--auth=none", "--proxy-path-passthrough=true"], "")
     const resp = await codeServer.fetch(proxyPath)
     expect(resp.status).toBe(200)
     const json = await resp.json()
     expect(json).toBe("joe is the best")
+  })
+
+  it("should rewrite redirects", async () => {
+    e.post("/wsup", (req, res) => {
+      res.redirect(307, "/finale")
+    })
+    e.post("/finale", (req, res) => {
+      res.json("redirect success")
+    })
+    ;[, , codeServer] = await integration.setup(["--auth=none"], "")
+    const resp = await codeServer.fetch(proxyPath, {
+      method: "POST",
+    })
+    expect(resp.status).toBe(200)
+    expect(await resp.json()).toBe("redirect success")
+  })
+
+  it("should not rewrite redirects", async () => {
+    const finalePath = proxyPath.replace("/wsup", "/finale")
+    e.post(proxyPath, (req, res) => {
+      res.redirect(307, finalePath)
+    })
+    e.post(finalePath, (req, res) => {
+      res.json("redirect success")
+    })
+    ;[, , codeServer] = await integration.setup(["--auth=none", "--proxy-path-passthrough=true"], "")
+    const resp = await codeServer.fetch(proxyPath, {
+      method: "POST",
+    })
+    expect(resp.status).toBe(200)
+    expect(await resp.json()).toBe("redirect success")
+  })
+
+  it("should allow post bodies", async () => {
+    e.use(bodyParser.json({ strict: false }))
+    e.post("/wsup", (req, res) => {
+      res.json(req.body)
+    })
+    ;[, , codeServer] = await integration.setup(["--auth=none"], "")
+    const resp = await codeServer.fetch(proxyPath, {
+      method: "post",
+      body: JSON.stringify("coder is the best"),
+      headers: {
+        "Content-Type": "application/json",
+      },
+    })
+    expect(resp.status).toBe(200)
+    expect(await resp.json()).toBe("coder is the best")
   })
 })

